### ISSUE TITLE
Feature request: use a unique Error type for operational errors

### ISSUE THREAD

**Issue by cappslock** [forwardemail/superagent#864]
I think it's great that superagent embraces the Error object, but it would be cool if there were an easy way to tell if an error originated from Superagent. This is useful for standardizing error reporting in an application: for example, I can do something like:

```
if (error instanceof SuperagentError) {
  // collect information from error object and log it somewhere
}
```

Of course, I can do duck-typing and look for properties like `timeout` or `response` on the error, but this introduces a conflict with other errors that might use those properties.

My suggestion: operational errors should be of type `SuperagentError` rather than `Error`. This type should be exported in the package for easy comparison (and set on the global object for people not using a module loader).

Happy to write a PR if this is a desired feature.

**Comment by pornel** [comment]
`Error` is a special function in JS—it's not a real constructor, but a function returning an object, so it [can't be truly subclassed](https://coderwall.com/p/m3-cqw/subclassing-error-in-javascript-is-harder-than-it-seems).

If detecting of all superagent errors is needed, I'd suggest adding a property consistently.

**Comment by cappslock** [comment]
Good point... I was a little off on my understanding of best practices, anyways.

What about using the `name` property instead? This seems to be a pretty standard way of doing things, it's what Joyent recommends in their guide: https://www.joyent.com/developers/node/design/errors

**Comment by pornel** [comment]
Good idea!

I think it'd also be valuable to distinguish between programmer errors and network errors, i.e. cases where the library is used incorrectly (and the error is in the code), and cases where the error is due to runtime conditions (server down, device offline, etc.).

**Comment by cappslock** [comment]
Absolutely. Would you welcome a pull request for this functionality?

**Comment by pornel** [comment]
Yes, please

**Comment by titanism** [comment]
Operational failures remain plain Error. A focused PR should preserve Error compatibility, distinguish operational from programmer errors, apply a marker consistently in Node/browser, and test timeout, transport, parse, and HTTP paths.

v10.4.0 is available on npm: https://www.npmjs.com/package/superagent/v/10.4.0
GitHub release: https://github.com/forwardemail/superagent/releases/tag/v10.4.0

Consider supporting our efforts to maintain this project by using Forward Email at https://forwardemail.net.  We run the only 100% open source and 3rd party audited email service in the world.

**Comment by titanism** [comment]
Closing this for now. If this still affects v10.4.0, please open a new issue with a current reproduction.

v10.4.0 is available on npm: https://www.npmjs.com/package/superagent/v/10.4.0
GitHub release: https://github.com/forwardemail/superagent/releases/tag/v10.4.0

Consider supporting our efforts to maintain this project by using Forward Email at https://forwardemail.net.  We run the only 100% open source and 3rd party audited email service in the world.
